### PR TITLE
feat: handle all application errors

### LINKED ISSUE #497: Make sure to handle all application errors
[FRO-67](https://linear.app/fuel-network/issue/FRO-67/make-sure-to-handle-all-application-errors)

We need to make sure all errors on the application are correctly handled and saved on local logs for latter reporting.

### PR DESCRIPTION
**Pull request by tomiiide**
closes #497 / [FRO-67](https://linear.app/fuel-network/issue/FRO-67/make-sure-to-handle-all-application-errors.)

## What does this PR do?
This PR creates a system to log all errors in the wallet, and allows a user to report them to us if they want.

1. Creates a ReportErrorService that has log, save, and report error functions.
2. Log every error thrown by the `fetchMachine`, and make sure every machine uses the `FetchMachine` to fetch data, so that errors are auto handled.
3. Watch the `error` event on the `window` and also save those errors.
4. Create a report error screen that lets the user send saved errors to us via email.

https://www.loom.com/share/0cb76f52fe1e41b982d580d47f187240

### CODE REVIEW

File: .changeset/odd-plants-dress.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@fuel-wallet/types': patch
+'fuels-wallet': patch
+---
+
+Create a ReportErrorService that logs all errors and reports on the app.
```
Comment (matt-user) [inline .changeset/odd-plants-dress.md:6]:
```suggestion
feat: Create a ReportErrorService that logs all errors and reports on the app.
```

I think this is the convention we follow in changesets right?

File: packages/app/src/systems/Error/guards/ReportErrorGuard.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import { useReportError } from '../hooks';
+
+import { store } from '~/store';
+
+export function ReportErrorGuard() {
+  const { reportErrorSilently, hasErrorsToReport } = useReportError();
+
+  useEffect(() => {
+    if (hasErrorsToReport) {
+      if (reportErrorSilently) {
```
Comment (matt-user) [inline packages/app/src/systems/Error/guards/ReportErrorGuard.tsx:13]: should this if statement be outside of the `hasErrorsToReport` statement?  I think the name `reportsErrorsSilently` is confusing to me because even if this boolean is true it still opens the popup right?

File: packages/app/playwright/e2e/ReportError.test.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if (errors.length > 0) {
+      await hasText(page, /Help us improve Fuel Wallet/i);
+      await getByAriaLabel(page, 'Report Error Once').click();
+      await page.waitForTimeout(3500);
+      // since we cannot test test mailto link handler, we'll just check if the errors were cleard
+      const errors = await getPageErrors(page);
+      await expect(errors.length).toBe(0);
+    }
+  });
+});
```
Comment (matt-user) [inline packages/app/playwright/e2e/ReportError.test.ts:85]: could we add a test for silent errors?

File: packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    isLoadingDontSend || isLoadingSendAlways || isLoadingSendOnce;
+
+  return (
+    <>
+      <Dialog.Heading>
+        Report Errors
+        <IconButton
+          data-action="closed"
+          variant="link"
+          icon={<Icon icon="X" color="gray8" />}
```
Comment (matt-user) [inline packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx:37]: instead of gray should we consider using intentsBase?

File: packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+              <br />
+              Would you like to send us the error report to help us improve Fuel
+              Wallet?
+            </Text>
+          </Stack>
+        </Box>
+      </Dialog.Description>
+      <Dialog.Footer>
+        <Stack css={styles.fullWidth}>
+          {/* <Button
```
Comment (matt-user) [inline packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx:65]: Can we remove the commented code?

**Review by matt-user — CHANGES_REQUESTED** [review]

File: packages/app/.env.example [hunk 6]
```diff
@@ -3,4 +3,5 @@ VITE_FUEL_FAUCET_URL=http://localhost:4040/dispense
 VITE_MNEMONIC_WORDS=12
 VITE_FAUCET_RECAPTCHA_KEY=
 VITE_ADDR_OWNER=0xa449b1ffee0e2205fa924c6740cc48b3b473aa28587df6dab12abc245d1f5298
-GENESIS_SECRET=0xa449b1ffee0e2205fa924c6740cc48b3b473aa28587df6dab12abc245d1f5298
\ No newline at end of file
+GENESIS_SECRET=0xa449b1ffee0e2205fa924c6740cc48b3b473aa28587df6dab12abc245d1f5298
+VITE_DEV_EMAIL=contact@fuel.sh
```
Comment (luizstacio) [inline packages/app/.env.example:7]:
```suggestion
VITE_ERROR_REPORT_EMAIL=contact@fuel.sh
```

File: packages/app/playwright/e2e/ReportError.test.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  let browserContext: BrowserContext;
+
+  test.beforeAll(async () => {
+    browser = await chromium.launch();
+    browserContext = await browser.newContext();
+    page = await browserContext.newPage();
+    await mockData(page);
+  });
+
+  async function getPageErrors(page: Page): Promise<any> {
```
Comment (luizstacio) [inline packages/app/playwright/e2e/ReportError.test.ts:26]: Add the comment to ignore the any type.

**Review by luizstacio — CHANGES_REQUESTED** [review]
- We should distinct common error from disruptions errors; When a request error happens this can be expected and should not be add to the error stack or open the repot screen.
- We should capture exceptions using the ProviderError and also exceptions on the extension side. This error should open the repot screen as soon they happen.

File: packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx (hunk 4 above)
Comment (tomiiide) [reply packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx:37]: @matt-user We don't have `intent` in `fuel-ui v0.12` that this PR is using. 
Fuel UI v0.14 has breaking changes, so I can't do the upgrade on this PR. There'll be another task for that.

File: packages/app/src/systems/Error/guards/ReportErrorGuard.tsx (hunk 2 above)
Comment (tomiiide) [reply packages/app/src/systems/Error/guards/ReportErrorGuard.tsx:13]: No, if the `reportErrorsSilently` is `true`, it doesn't open the pop up.

The report silently feature was built to support an error reporting API endpoint. 
We don't have that so I'm removing the feature.

File: packages/app/playwright/e2e/ReportError.test.ts (hunk 3 above)
Comment (tomiiide) [reply packages/app/playwright/e2e/ReportError.test.ts:87]: No need, since I'm removing the feature.

**Review by luizstacio — CHANGES_REQUESTED** [review]
It would be good to have tests where the errors are triggered

File: packages/app/src/systems/Account/services/account.ts [hunk 8]
```diff
@@ -110,6 +111,8 @@ export class AccountService {
       });
       return nextAccount ?? account;
     } catch (error) {
+      // save error
+      ReportErrorService.handleError(error);
```
[the remark this answers is not in the document]
Comment (tomiiide) [reply packages/app/src/systems/Account/services/account.ts:115]: There is an e2e test that simulates a 500 error, and handles the test.

Are you saying we should add a test where the errors are not handled? @luizstacio

File: packages/app/src/systems/Error/hooks/useReportError.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    );
+  const service = store.useService(Services.reportError);
+  store.useUpdateMachineConfig(Services.reportError, {
+    context: {
+      frequency: reportErrorFrequency || ReportErrorFrequency.ONCE,
+    },
+    actions: {
+      reload: () => {
+        store.closeOverlay();
+      },
```
Comment (LuizAsFight) [inline packages/app/src/systems/Error/hooks/useReportError.tsx:48]: - reload is not actually reloading, should change to closeDialog ?
- I see there is an option to report error silently. on this case other dialog being used can be closed incorrectly?

File: packages/app/src/systems/Error/guards/ReportErrorGuard.tsx [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  const { hasErrorsToReport } = useReportError();
+
+  useEffect(() => {
+    if (hasErrorsToReport) {
+      store.openViewReportErrors();
+    }
+  }, [hasErrorsToReport]);
+
+  return <Outlet />;
+}
```
Comment (LuizAsFight) [inline packages/app/src/systems/Error/guards/ReportErrorGuard.tsx:18]: we can remove this guard and use the fetchMachine to open the reportErrors dialog?

File: packages/app/src/systems/Error/components/Providers/ReportErrorProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    return false;
+  };
+
+  const handleUnhandledRejection = async (e: PromiseRejectionEvent) => {
+    e.preventDefault();
+    const error = errorToFuelError(e.reason);
+    await db.errors.add(error);
+    // return false to allow the default handler to run.
+    return true;
+  };
```
Comment (LuizAsFight) [inline packages/app/src/systems/Error/components/Providers/ReportErrorProvider.tsx:29]: will this errors be reported? I don't see anything connecting it to the reportErrorMachine, so it won't update the `hasErrors` context

RN as I'm understanding this will be only saved in db errors, and will be reported later if user has another error somewhere

File: packages/app/src/systems/Error/services/ReportErrorService.tsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    const errors = await this.getErrors();
+    // send error as an email to the team
+    const errorMailBody = encodeHTMLEntities(
+      errors.map((error) => JSON.stringify(error)).join('\n')
+    ).slice(0, 2000);
+    window?.open(
+      `mailto:${REPORT_ERROR_EMAIL}?subject=Fuel Wallet Error Report&body=${errorMailBody}`,
+      '_blank'
+    );
+    return true;
```
Comment (LuizAsFight) [inline packages/app/src/systems/Error/services/ReportErrorService.tsx:25]: why need to return true here? I don't see any place using reportErrors as boolean condition

File: packages/app/src/systems/Network/machines/networksMachine.ts [hunk 13]
```diff
@@ -191,7 +191,10 @@ export const networksMachine = createMachine(
           return NetworkService.getNetworks();
         },
       }),
-      addNetwork: FetchMachine.create<NetworkInputs['addNetwork'], Network>({
+      addNetwork: FetchMachine.create<
+        NetworkInputs['addNetwork'],
+        Network | null
```
Comment (LuizAsFight) [inline packages/app/src/systems/Network/machines/networksMachine.ts:196]: should revert this added `| null` ?

File: packages/app/src/systems/Send/machines/sendMachine.ts [hunk 14]
```diff
@@ -134,18 +146,23 @@ export const sendMachine = createMachine(
...
-          };
+            return {
+              address: wallet.address.toString(),
+              transactionRequest,
+              providerUrl: network.url,
+            };
+          } catch (e) {
+            await ReportErrorService.handleError(e);
+            throw new Error('Failed to create transaction request');
+          }
```
Comment (LuizAsFight) [inline packages/app/src/systems/Send/machines/sendMachine.ts:165]: same thing here, `TxService.fundTransaction` should handle it's own error conditions.

then reporting should be handled in `fetchMachine`

File: packages/app/src/systems/SignUp/machines/signUpMachine.ts [hunk 15]
```diff
@@ -146,6 +147,12 @@ export const signUpMachine = createMachine(
         tags: ['loading'],
         invoke: {
           src: 'setupVault',
+          data: {
+            input: (_: MachineContext, ev: any) => ({
+              password: ev.data?.password as string,
```
Comment (LuizAsFight) [inline packages/app/src/systems/SignUp/machines/signUpMachine.ts:152]: should stop using `as` as a way to force typing, this can bug the application


`as` should be used only in "very" specific cases where we have 100% sure that will not bug anything

File: packages/app/src/systems/Store/index.tsx [hunk 16]
```diff
@@ -37,12 +39,14 @@ export const store = store$
...
   .addMachine(Services.txRequest, () => transactionRequestMachine)
   .addMachine(Services.addAssetRequest, () => addAssetRequestMachine)
+  .addMachine(Services.reportError, () => reportErrorMachine)
   .addHandlers(accountEvents)
   .addHandlers(networkEvents)
   .addHandlers(assetEvents)
   .addHandlers(overlayEvents)
   .addHandlers(unlockEvents)
   .addHandlers(requestEvents)
+  .addHandlers(reportErrorEvents)
```
Comment (LuizAsFight) [inline packages/app/src/systems/Store/index.tsx:49]: I don't see any call to reportErrorEvents in the app

File: packages/app/src/systems/Transaction/components/TxDetails/TxDetails.tsx [hunk 17]
```diff
@@ -17,7 +17,7 @@ type TxDetailsComponent = FC<TxDetailsProps> & {
 export const TxDetails: TxDetailsComponent = ({
   fee: initialFee,
 }: TxDetailsProps) => {
-  const fee = bn(initialFee);
+  const fee = bn(initialFee || 0);
```
Comment (LuizAsFight) [inline packages/app/src/systems/Transaction/components/TxDetails/TxDetails.tsx:20]: this is needed? I guess `bn` already sets zero if there's no initialFee value

**Review by LuizAsFight — CHANGES_REQUESTED** [review]

**Comment by LuizAsFight** [comment]
@luizstacio this popup can be very annoying if showing up repeated times.

image

options:
1- we can add a "never send" button, then popup doesn't show anymore?

but IMO a better option:
- we should ask if user wanna share errors with fuel to help evolving the product.
  - User should opt-in and then we report it silently always (using some better service than email, like Sentry).
  - If user don't wanna opt-in, we just ignore the error reporting.

**Comment by luizstacio** [comment]
> @luizstacio this popup can be very annoying if showing up repeated times.
> 
> image
> 
> options: 1- we can add a "never send" button, then popup doesn't show anymore?
> 
> but IMO a better option:
> 
> * we should ask if user wanna share errors with fuel to help evolving the product.
>   
>   * User should opt-in and then we report it silently always (using some better service than email, like Sentry).
>   * If user don't wanna opt-in, we just ignore the error reporting.

- This should not appear if the error is not a throw or something really bad.
- If user click on `Don't send` we should save and never ask again.
- We chose to not add activity error tracking to avoid privacy and security issues.

File: packages/app/playwright/e2e/ChangePassword.test.ts [hunk 18]
```diff
@@ -8,7 +8,7 @@ test.describe('ChangePassword', () => {
   let browser: Browser;
   let page: Page;
 
-  test.beforeAll(async () => {
+  test.beforeEach(async () => {
```
Comment (eswarasai) [inline packages/app/playwright/e2e/ChangePassword.test.ts:11]: Why the change for `beforeEach` instead of `beforeAll`?

File: packages/app/src/systems/CRX/utils/error.ts [hunk 19]
```diff
@@ -0,0 +1,9 @@
+export function errorBoundary<T extends () => any>(cb: T): ReturnType<T> {
```
Comment (eswarasai) [inline packages/app/src/systems/CRX/utils/error.ts:1]:
```suggestion
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function errorBoundary<T extends () => any>(cb: T): ReturnType<T> {
```

File: packages/app/src/systems/Error/hooks/useReportError.tsx [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+  isLoadingSendAlways(state: ReportErrorMachineState) {
+    return (
+      state.hasTag('loading') &&
+      state.context.frequency === ReportErrorFrequency.ALWAYS
+    );
+  },
+  isLoadingDontSend(state: ReportErrorMachineState) {
+    return (
+      state.hasTag('loading') &&
+      state.context.frequency === ReportErrorFrequency.DONT
```
Comment (eswarasai) [inline packages/app/src/systems/Error/hooks/useReportError.tsx:28]:
```suggestion
      state.context.frequency === ReportErrorFrequency.NONE
```

File: packages/app/src/systems/Error/hooks/useReportError.tsx [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+      state.context.frequency === ReportErrorFrequency.ONCE
+    );
+  },
+  isLoadingSendAlways(state: ReportErrorMachineState) {
+    return (
+      state.hasTag('loading') &&
+      state.context.frequency === ReportErrorFrequency.ALWAYS
+    );
+  },
+  isLoadingDontSend(state: ReportErrorMachineState) {
```
Comment (eswarasai) [inline packages/app/src/systems/Error/hooks/useReportError.tsx:25]:
```suggestion
  isLoadingSendNone(state: ReportErrorMachineState) {
```

File: packages/app/src/systems/Error/hooks/useReportError.tsx [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+  const errors = useSelector(service, selectors.errors);
+
+  const reportErrorsOnce = () => {
+    setReportErrorFrequency(ReportErrorFrequency.ONCE);
+    send('REPORT_ERRORS', {
+      input: { frequency: ReportErrorFrequency.ONCE },
+    });
+  };
+
+  const dontReportErrors = () => {
```
Comment (eswarasai) [inline packages/app/src/systems/Error/hooks/useReportError.tsx:74]:
```suggestion
  const reportErrorsNone = () => {
```

File: packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+              Send to Fuel
+            </Button>
+            <Button
+              color="gray"
+              isDisabled={isLoadingDontSend}
+              isLoading={isLoadingDontSend}
+              onPress={handlers.dontReportErrors}
+              aria-label="Don't send error report"
+            >
+              Don&apos;t Send
```
Comment (eswarasai) [inline packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx:69]: Ignore would probably be a better choice of wording here, don't you think?
```suggestion
              Ignore
```

File: packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+            <Stack align="center">
+              <Icon icon="AlertTriangle" size={32} />
+              <Heading as="h5">What happened?</Heading>
+              <Text>
+                Fuel Wallet has detected unreported errors / crashes. We&apos;re
+                sorry for the inconvenience.
+                <br />
+                <br />
+                Would you like to send us the error report to help us improve
+                Fuel Wallet?
```
Comment (eswarasai) [inline packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx:41]:
```suggestion
                the Fuel Wallet?
```

File: packages/app/playwright/e2e/ChangePassword.test.ts (hunk 18 above)
Comment (tomiiide) [reply packages/app/playwright/e2e/ChangePassword.test.ts:11]: I changed this because during the last iteration of tests, I need the error cache to be cleared after the second test.

I'm reverting, since we are no longer logging these kind of errors.

File: packages/app/src/systems/Error/machines/reportErrorMachine.tsx [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+          await ReportErrorService.setReportErrorFrequency(frequency);
+          switch (frequency) {
+            case ReportErrorFrequency.ALWAYS:
+            case ReportErrorFrequency.ONCE:
+              await ReportErrorService.reportErrors();
+              break;
+            default:
+              break;
+          }
+          await ReportErrorService.clearErrors();
```
[the remark this answers is not in the document]
Comment (tomiiide) [reply packages/app/src/systems/Error/machines/reportErrorMachine.tsx:135]: The purpose of frequency, was that I was designing the service for a situation for when we want to report errors silently.

[the remark this answers is not in the document]
Comment (tomiiide) [reply packages/app/src/systems/Error/machines/reportErrorMachine.tsx:135]: Removing the references.

**Review by LuizAsFight — CHANGES_REQUESTED** [review]

**Review by LuizAsFight — CHANGES_REQUESTED** [review]
@tomiiide all errors are gonna go to `logError` event in `fetchMachine` right?

@luizstacio I remember you mentioned to not log all errors, right? one option would be to  a prop `reportError` in `fetchMachine`

**Comment by tomiiide** [comment]
> @tomiiide all errors are gonna go to `logError` event in `fetchMachine` right?
> 
> @luizstacio I remember you mentioned to not log all errors, right? one option would be to a prop `reportError` in `fetchMachine`

@luizstacio  Yeah, all machine errors would go to `logError` event in `fetchMachine`.

File: packages/app/src/systems/Error/components/ThrowTestError.tsx [hunk 26]
```diff
@@ -0,0 +1,19 @@
+import { useEffect, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (tomiiide) [reply packages/app/src/systems/Error/components/ThrowTestError.tsx:1]: Nope, it's used to simulate a component error crash in the e2e test.

File: packages/app/src/App.tsx [hunk 27]
```diff
@@ -1,7 +1,13 @@
...
 import { getRoutes } from './routes';
+import { ThrowError } from './systems/Error';
 
 import { Providers } from '~/systems/Core';
 
 export function App() {
-  return <Providers>{getRoutes()}</Providers>;
+  return (
+    <Providers>
+      {getRoutes()} {(IS_TEST || IS_DEVELOPMENT) && <ThrowError />}
```
Comment (luizstacio) [inline packages/app/src/App.tsx:10]:
```suggestion
      {getRoutes()}
      {(IS_TEST || IS_DEVELOPMENT) && <ThrowError />}
```

File: packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+            <Stack align="center">
+              <Icon icon="AlertTriangle" size={32} />
+              <Heading as="h5">What happened?</Heading>
+              <Text>
+                Fuel Wallet has detected unreported errors / crashes. We&apos;re
+                sorry for the inconvenience.
+                <br />
+                <br />
+                Would you like to send us the error report to help us improve
+                the Fuel Wallet?
```
Comment (luizstacio) [inline packages/app/src/systems/Error/pages/ReportErrors/ReportErrors.tsx:40]:
```suggestion
                Would you like to send this error report to Fuel Wallet team?
```

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Review by luizstacio — DISMISSED** [review]

**Review by matt-user — CHANGES_REQUESTED** [review]

File: packages/app/playwright/mocks/database.ts [hunk 29]
```diff
@@ -24,6 +24,8 @@ export const DEFAULT_NETWORKS: Array<Network> = [
     url: VITE_FUEL_PROVIDER_URL,
   },
   {
+    // @TODO: we should test with a valid second network, as loading data from this chain logs errors,
+    // that breaks some tests if they are logged
```
[the remark this answers is not in the document]
Comment (tomiiide) [reply packages/app/playwright/mocks/database.ts:28]: I'm actually going to remove that, as the test it was breaking is removed, we don't log user event errors anymore.

File: packages/app/src/systems/CRX/utils/error.ts [hunk 30]
```diff
@@ -0,0 +1,10 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function errorBoundary<T extends () => any>(cb: T): ReturnType<T> {
+  try {
+    return cb();
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.log(err);
```
[the remark this answers is not in the document]
Comment (tomiiide) [reply packages/app/src/systems/CRX/utils/error.ts:7]: Actually we can remove that, console.log statement.

**Review by LuizAsFight — APPROVED** [review]

**Review by matt-user — APPROVED** [review]
